zkevm-circuits/src/evm_circuit/execution/memory_copy.rs

```diff
@@ -0,0 +1,532 @@
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::CopyToMemory;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let src_addr = cb.query_cell();
+        let dst_addr = cb.query_cell();
+        let bytes_left = cb.query_cell();
+        let src_addr_end = cb.query_cell();
+        let from_tx = cb.query_bool();
+        let tx_id = cb.query_cell();
+        let buffer_get_data =
+            BufferGetDataGadget::construct(cb, &src_addr, &src_addr_end);
+
+        let from_memory = 1.expr() - from_tx.expr();
+        let rw_counter = cb.curr.state.rw_counter.expr();
+        let mut rw_counter_offset = 0.expr();
+        // Copy bytes from src and dst
+        for i in 0..MAX_COPY_BYTES {
+            let read_flag = buffer_get_data.read_from_buffer(i);
+            // Read bytes[i] from memory
+            cb.condition(from_memory.clone() * read_flag.clone(), |cb| {
+                cb.memory_lookup_with_counter(
+                    rw_counter.clone() + rw_counter_offset.clone(),
+                    0.expr(),
+                    src_addr.expr() + i.expr(),
+                    buffer_get_data.byte(i),
+                )
+            });
+            rw_counter_offset =
+                rw_counter_offset + from_memory.clone() * read_flag.clone();
+            // Read bytes[i] from Tx
+            cb.condition(from_tx.expr() * read_flag.clone(), |cb| {
+                cb.tx_lookup(
+                    tx_id.expr(),
+                    TxContextFieldTag::CallData,
+                    Some(src_addr.expr() + i.expr()),
+                    buffer_get_data.byte(i),
+                )
+            });
+            // Write bytes[i] to memory when selectors[i] != 0
+            cb.condition(buffer_get_data.has_data(i), |cb| {
+                cb.memory_lookup_with_counter(
+                    rw_counter.clone() + rw_counter_offset.clone(),
+                    1.expr(),
+                    dst_addr.expr() + i.expr(),
+                    buffer_get_data.byte(i),
+                )
+            });
+            rw_counter_offset = rw_counter_offset + buffer_get_data.has_data(i);
+        }
+
+        let copied_size = buffer_get_data.num_bytes();
+        let finish_gadget = ComparisonGadget::construct(
+            cb,
+            copied_size.clone(),
+            bytes_left.expr(),
+        );
+        let (lt, finished) = finish_gadget.expr();
+        cb.add_constraint(
+            "Constrain num_bytes <= bytes_left",
+            lt * finished.clone(),
+        );
+
+        // When finished == 0, constraint the CopyToMemory state in next step
+        let next_src_addr = cb.query_cell_next_step();
+        let next_dst_addr = cb.query_cell_next_step();
+        let next_bytes_left = cb.query_cell_next_step();
+        let next_src_addr_end = cb.query_cell_next_step();
+        let next_from_tx = cb.query_cell_next_step();
+        let next_tx_id = cb.query_cell_next_step();
```

Comment:
ok. If i make it right, the order of these cells of next step and current step, must be exact same. (src_addr, dst_addr, ...). or else contraints will fail.      So maybe it is better to add a function to cb called 'query_current_and_next_cell_pair() -> (Cell, Cell)' rather than use query_cell_next_step seperately. It is very easy to incur bugs in the future.

Reply:
I added comments to the `query_cell_next_step`. Could you take a look?